runtime/compiler/wasm/reader_test.go

```diff
@@ -0,0 +1,825 @@
+			InvalidFuncTypeResultTypeError{
+				Index: 0,
+				ReadError: InvalidValTypeError{
+					Offset:  9,
+					ValType: 0xFF,
+				},
+			},
+			err,
+		)
+		assert.Nil(t, funcTypes)
+	})
+
+	t.Run("invalid, result type missing", func(t *testing.T) {
+
+		t.Parallel()
+
+		funcTypes, err := read([]byte{
+			// section size: 4 (LEB128)
+			0x85, 0x80, 0x80, 0x80, 0x0,
+			// type count
+			0x1,
+			// function type
+			0x60,
+			// parameter count: 0
+			0x0,
+			// result count: 2
+			0x2,
+			// type of result 1: i32
+			0x7f,
+		})
+		require.Error(t, err)
+		assert.Equal(t,
+			InvalidFuncTypeResultTypeError{
+				Index: 1,
+				ReadError: InvalidValTypeError{
+					Offset:    10,
+					ValType:   0,
+					ReadError: io.EOF,
+				},
+			},
+			err,
+		)
+		assert.Nil(t, funcTypes)
+	})
+}
+
+func TestWasmReader_readFunctionSection(t *testing.T) {
+
+	t.Parallel()
+
+	read := func(data []byte) ([]uint32, error) {
+		b := buf{data: data}
+		r := WASMReader{buf: &b}
+		err := r.readFunctionSection()
+		if err != nil {
+			return nil, err
+		}
+		return r.Module.functionTypeIDs, nil
+	}
+
+	t.Run("valid", func(t *testing.T) {
+
+		t.Parallel()
+
+		typeIDs, err := read([]byte{
+			// section size: 2 (LEB128)
+			0x82, 0x80, 0x80, 0x80, 0x0,
+			// function count: 1
+			0x1,
+			// type ID of function: 0x23
+			0x23,
+		})
+		require.NoError(t, err)
+		assert.Equal(t,
+			[]uint32{0x23},
+			typeIDs,
+		)
+	})
+
+	t.Run("invalid size", func(t *testing.T) {
+
+		t.Parallel()
+
+		funcTypes, err := read([]byte{
+			0x87, 0x80, 0x80,
+		})
+		require.Error(t, err)
+		assert.Equal(t,
+			InvalidSectionSizeError{
+				Offset:    0,
+				ReadError: io.EOF,
+			},
+			err,
+		)
+		assert.Nil(t, funcTypes)
+	})
+
+	t.Run("invalid count", func(t *testing.T) {
+
+		t.Parallel()
+
+		funcTypes, err := read([]byte{
+			// section size: 7 (LEB128)
+			0x80, 0x80, 0x80, 0x80, 0x0,
+		})
+		require.Error(t, err)
+		assert.Equal(t,
+			InvalidFunctionSectionFunctionCountError{
+				Offset:    5,
+				ReadError: io.EOF,
+			},
+			err,
+		)
+		assert.Nil(t, funcTypes)
+	})
+
+	t.Run("invalid function type ID", func(t *testing.T) {
+
+		t.Parallel()
+
+		funcTypes, err := read([]byte{
+			// section size: 7 (LEB128)
+			0x81, 0x80, 0x80, 0x80, 0x0,
+			// function count
+			0x1,
+		})
+		require.Error(t, err)
+		assert.Equal(t,
+			InvalidFunctionSectionFunctionTypeIDError{
+				Offset:    6,
+				Index:     0,
+				ReadError: io.EOF,
+			},
+			err,
+		)
+		assert.Nil(t, funcTypes)
+	})
+}
+
+func TestWasmReader_readCodeSection(t *testing.T) {
+
+	t.Parallel()
+
+	read := func(data []byte) ([]*Code, error) {
+		b := buf{data: data}
+		r := WASMReader{buf: &b}
+		err := r.readCodeSection()
+		if err != nil {
+			return nil, err
+		}
+		return r.Module.functionBodies, nil
+	}
+
+	t.Run("valid", func(t *testing.T) {
+
+		t.Parallel()
+
+		codes, err := read([]byte{
+			// section size: 15 (LEB128)
+			0x8f, 0x80, 0x80, 0x80, 0x0,
+			// function count: 1
+			0x1,
```

Comment:
I was afraid these raw bytes would be unreadable, and useless for review. 
But after reading the code, these bytes and comments are actually quite easier to read!
Nice comments 👍

Reply:
I'm glad they're helpful! I think once you have internalized the structure of the binary format, it becomes readable and when something is mis-produced, it helps with debugging.